apiserver/facades/controller/sshserver/shim.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshserver
+
+import (
+	"strconv"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/core/virtualhostname"
+	"github.com/juju/juju/state"
+)
+
+type backend struct {
+	*state.StatePool
+}
+
+// ControllerConfig gets the controller config from the systemState.
+func (b backend) ControllerConfig() (controller.Config, error) {
+	systemState, err := b.StatePool.SystemState()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return systemState.ControllerConfig()
+}
+
+// ControllerConfig gets the ssh server hostkey from the systemState.
+func (b backend) SSHServerHostKey() (string, error) {
+	systemState, err := b.StatePool.SystemState()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return systemState.SSHServerHostKey()
+}
+
+// WatchControllerConfig gets the controller config watcher from the systemState.
+func (b backend) WatchControllerConfig() (state.NotifyWatcher, error) {
+	systemState, err := b.StatePool.SystemState()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return systemState.WatchControllerConfig(), nil
+}
+
+func (b backend) HostKeyForVirtualHostname(info virtualhostname.Info) ([]byte, error) {
+	model, poolHelper, err := b.StatePool.GetModel(info.ModelUUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer poolHelper.Release()
+	stateModel := model.State()
+	switch info.Target() {
+	case virtualhostname.MachineTarget:
+		machineId, _ := info.Machine()
+		hostkey, err := stateModel.MachineVirtualHostKey(strconv.Itoa(machineId))
+		if err != nil {
+			return nil, err
+		}
+		return hostkey.HostKey(), nil
+	case virtualhostname.ContainerTarget, virtualhostname.UnitTarget:
+		unitName, _ := info.Unit()
+		hostkey, err := stateModel.UnitVirtualHostKey(unitName)
+		if err != nil {
+			return nil, errors.Annotate(err, "failed to get unit host key")
+		}
+		return hostkey.HostKey(), nil
+	default:
+		return nil, errors.NotValidf("unsupported target: %v", info.Target())
+	}
```

Comment:
This shim is doing a lot of business logic that should be done outside this (so you can test it appropriately).